pkg/actor/validate_version.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2021 The Cockroach Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actor
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	api "github.com/cockroachdb/cockroach-operator/api/v1alpha1"
+	"github.com/cockroachdb/cockroach-operator/pkg/condition"
+	"github.com/cockroachdb/cockroach-operator/pkg/kube"
+	"github.com/cockroachdb/cockroach-operator/pkg/ptr"
+	"github.com/cockroachdb/cockroach-operator/pkg/resource"
+	"github.com/pkg/errors"
+	kbatch "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/cockroachdb/cockroach-operator/pkg/features"
+	"github.com/cockroachdb/cockroach-operator/pkg/utilfeature"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"github.com/cenkalti/backoff"
+	"github.com/go-logr/logr"
+)
+
+func newVersionChecker(scheme *runtime.Scheme, cl client.Client, config *rest.Config) Actor {
+	return &versionChecker{
+		action: newAction("Crdb Version Validator", scheme, cl),
+		config: config,
+	}
+}
+
+// validate performs the validation of the crdb image for the new cluster
+type versionChecker struct {
+	action
+
+	config *rest.Config
+}
+
+func (v versionChecker) GetActionType() api.ActionType {
+	return api.VersionCheckerAction
+}
+
+func (v versionChecker) Handles(conds []api.ClusterCondition) bool {
+	return utilfeature.DefaultMutableFeatureGate.Enabled(features.CrdbVersionValidator) && (condition.False(api.NotInitializedCondition, conds) || condition.True(api.NotInitializedCondition, conds)) && condition.True(api.CrdbVersionNotChecked, conds)
+}
+
+func (v versionChecker) Act(ctx context.Context, cluster *resource.Cluster) error {
```

Comment:
This is a very long func.  We need to break it into a few smaller funcs that can be unit test easier.  This is future work